ouroboros-consensus/docs/VersioningSchemeDecision.md

```diff
@@ -0,0 +1,286 @@
+  We'd like it to require as few inputs, discussions, decisions, etc as possible.
+
+- *Distinguished Development Versions*.
+  We'd like to distinguish between the two possible semantics of a version number.
+
+    - The version of a released thing identifies some immutable thing, which is always somehow more refined than any thing---but especially another _released_ thing---that has a lesser version number.
+      We denote this kind of version by "release version" below.
+
+    - A development version refers to some mutable thing that is improving during the time between two releases, eg the version on the main integration branch.
+      Note in particular that there will usually be multiple different commits that all have the same development version number.
+      We denote this kind of version by "`main` version" below.
+
+The main integration branch is typically named `main` in fresh GitHub repositories, so that's what we'll use in this document.
+
+## Proposal RisingEdge
+
+PRs do not alter the `main` version.
+
+To cut a release from a commit COMMIT1 on `main` that declares version `A.B.C`, add to `main` a commit COMMIT2 that extends COMMIT1 merely to declare the version `next(A.B).0` or `A.B.(C+1)` depending on what has changed on `main` since the previous release, and announce COMMIT2.
+
+Cons:
+
+- `main` versions and release versions are not distinguished.
+
+- Cutting a release requires assessing all the changes on `main` since the last release.
+
+## Proposal FallingEdgePatch
+
+A `C`-level PR doesn't alter the `main` version.
+Each `A`-level PR must update the `main` version from `A.B.C` to `(A+1).B.0` unless `A.0.0` is already greater than the previous release.
+Each `B`-level PR must update the `main` version from `A.B.C` to `A.(B+1).0` unless `A.B.0` is already greater than the previous release.
+
+To cut a release from a commit COMMIT on `main` that declares `A.B.C`, announce COMMIT.
+Also immediately update the `main` version to `A.B.(C+1)`.
+
+Cons:
+
+- `main` versions and release versions are not distinguished.
+
+- Some `main` commits will declare version `A.B.(C+1)` even if that version is never officially released or is created only later by backporting _different_ (patch) commits to a previous release branch.
+
+## Proposal Parity
+
+Each `main` version `A.B.C` has an odd `B`.
+Each release version `A.B.C` has an even `B`.
+This is similar to the GHC Team's scheme.
+
+PRs do not alter the `main` version.
+
+To cut a release from a commit COMMIT on `main` that declares version `A.B` (where `B` is necessarily odd), announce a new non-`main` commit that extends COMMIT merely to declare version `X.Y = next(A.B)` (where `Y` is necessarily even).
+Also immediately update the `main` version to `X.(Y+1)`.
+
+Cons:
+
+- Commits on `main` that only include patches since the previous release would still spuriously include the `Y+1` increment in the `B` dimension.
```

Comment:
I wonder in which cases this will be an actual problem. Also, if we bumped `Y` on `main`, making the current version `X.(Y+1).Z`, I would expect that commits that include only patches would bump `Z` (which I understand it is not what is proposed here).

Reply:
I think because:

- It's confusing; people tend to expect versions to evolve in a minimal way.
- Spurious increases in bounds cause unnecessary busy work for downstream .cabal files that use (relatively) strong upper bounds on our libraries.

Reply:
Observation: so a new minor release `A.B.(C+1)` (with `B` even) will only be possible from a release branch hooked on `A.B.0`.  This is compatible with the branching model we use right now.